maze-consulting/SFCC_Klaviyo_Maze
# Feature/kscu 39 extract emails forms

## Description

This update handles a new feature that applies listeners to inputs on any page to extract key identifying information from forms anywhere on the site _(both SFRA & SiteGen)_. This new Klaviyo listener captures emails & phone numbers a user enters in a form & sends that to `klaviyo.identify()` to identify a user _(ex: Adds the user's email and/or email to the Profiles page in the Klaviyo dashboard and creates an exchange ID)_. 

Inputs are selected based on CSS selectors entered into a new site preference in Business Manager _(Pref ID: `klaviyo_email_selectors`)_.  Multiple tickets are addressed in this update. 


## Manual Testing Steps

This was tested in both SFRA & SiteGen across multiple browsers _(Chrome, FireFox, Safari)_. Manual tests included: 

- Inspecting the DOM after entering CSS selectors into the 'klaviyo_email_selectors' Site Preference in BM to confirm targeted inputs received applied attributes _(ex: `data-listener="klaviyo"`)_
- Clearing cookies & entering **valid** test email addresses into email fields - checking console values & Klaviyo dashboard for new profiles
- Clearing cookies & entering **invalid** test email addresses & other values to check for errors & confirm no new profiles were created in the Klaviyo dashboard
- Clearing cookies & entering test phone numbers into phone fields to confirm invalid values were not sent to the `klaviyo.idenfity()` call
- Clearing cookies & entering valid test phone numbers - checking console values & Klaviyo dashboard for newly created profiles
- Creating a button to render new inputs to the screen to mimick pop-ups AFTER the DOM was fully loaded & confirming inputs with applied listeners were functioning as expected


## Additional Note: Site Preference

This update requires a new Site Preference where Business Users can enter selectors. The meta data with the new site pref is available within the `custom-preference_KSCU-39.xml` file.
